---
title: "Visualisation"
author: "Philipp"
date: "1/11/2019"
output:
  bookdown::html_document2:
    toc: true
    theme: darkly
---

# Introduction to ggplot2

Learning how to use the ggplot2 package to visualize sequence data. 

```{r Setup, echo=FALSE}
library(ggplot2)
```

```{r Read&setup_data, echo=FALSE, include=FALSE}
variantData <- read.delim("http://genomedata.org/gen-viz-workshop/intro_to_ggplot2/ggplot2ExampleData.tsv")
head(variantData)
variantData[1:5, c(1:9,16,18)]
variantData$tumor_COV <- variantData$tumor_ref_count + variantData$tumor_var_count
```

## basics

The following results in the same graphic, however, be aware in figure p2 that you have to follow this order, or at least the `geom_point()` argument with the `aes()` in the end, otherwise the plot kinda does not work.

```{r plot1&2, echo=TRUE, fig.cap="Figure 1&2: Comparison of two ways to create the same plot"}
p1 <- ggplot(data = variantData, aes(x = tumor_VAF, y = tumor_COV))
p1 <- p1 + geom_point()
p1 <- p1 + theme_classic()
p1 <- p1 + ggtitle('p1')
p1

p2 <- ggplot()+
  theme_classic()+
  ggtitle('p2')+
  geom_point(data = variantData, aes(x = tumor_VAF, y = tumor_COV))
p2
```

Now we are going to spead out the data a little more...

```{r plot3, echo=FALSE, fig.cap="Figure 3: Jitter"}
p3 <- ggplot()+
  geom_point(data = variantData, aes(x = chromosome_name, y = tumor_VAF), position = 'jitter')
p3
```

Limting the axis of the plots

```{r plot4, echo=FALSE, fig.cap="Figure 4: limiting the axis"}
p4 <- p1 + scale_y_continuous(limits = c(0,500))
p4
```

```{r plot5, echo=FALSE, fig.cap="Figure 5: transforming the actual values in the data frame"}
p5 <- ggplot()+
  theme_classic()+
  geom_point(data = variantData, aes(x = tumor_VAF, y = log2(tumor_COV)), size = .7, shape = 3)
p5
```

```{r plot6, echo=FALSE, fig.cap="Figure 6: transform the axis using ggplot"}
p6 <- p1 + scale_y_continuous(trans = 'log2') + ylab('log2(tumor_COV)')
p6
```

## Applying different aesthetics

```{r plot7, echo=FALSE, fig.cap="Figure 7: what happens when we try to add color within the aesthetics?"}
p7 <- ggplot()+
  geom_point(data = variantData, aes(x = tumor_VAF, y = tumor_COV, color = 'blue'))+
  scale_y_continuous(trans = 'log2')
p7
```

```{r plot8, echo=FALSE, fig.cap="Figure 8: and what happens when we try to add color within the geom"}
p8 <- ggplot()+
  geom_point(data = variantData, aes(x = tumor_VAF, y = tumor_COV), color = 'blue')+
  scale_y_continuous(trans = 'log2')
p8
```

```{r plot9, echo=FALSE, fig.cap="Figure 9: get a density curve of tumor vafs"}
p9 <- ggplot()+
  geom_density(data = variantData, aes(x = tumor_VAF, color = dataset))
p9
```

`adjust =` in the `geom_density` allows for smoothing of the density function.

```{r plot10, echo=FALSE, fig.cap="Figure 10: let's add a bit more detail"}
p10 <- ggplot()+
  geom_density(data = variantData, aes(x = tumor_VAF, fill = dataset), alpha = .5, color = 'black', adjust = .5)
p10
```

```{r plot11, echo=FALSE, fig.cap="Figure 11: and let's change the colors some more"}
p11 <- p10+
  scale_fill_brewer(palette = 'Set1')
p11
```

```{r plot12, echo=FALSE, fig.cap="Figure 12: putting different colors..."}
p12 <- ggplot()+
  geom_density(data = variantData, aes(x = tumor_VAF, fill = dataset, color = dataset))+
  scale_fill_brewer(palette = 'Greys')+
  theme_classic()
p12
```

## Faceting

```{r plot13, echo=FALSE, fig.cap="Figure 13: relation of tiers to mutation type"}
p13 <- ggplot(data = variantData[variantData$type == 'SNP',])+
  geom_bar(aes(x = trv_type, fill = tier))+
  scale_fill_brewer(palette = 'Set1')+
  theme_classic()
p13
```

```{r plot14, echo=FALSE, fig.cap="Figure 14: which reference base is most often mutated"}
p14 <- p13 + facet_wrap(~reference)
p14
```

```{r plot15, echo=FALSE, fig.cap="Figure 15: which transition and transversions occur mist frequently"}
p15 <- p14 +
  facet_grid(variant~reference)+
  theme(axis.text.x = element_text(angle = 90, size = 3))
p15
```

Selecting form the `data.frame`

```{r selecting, echo=TRUE}
x <- variantData[variantData$type=='SNP',]
dim(x)
dim(variantData)
#head(x)
```

## ggplot themes

```{r plot16, echo=FALSE, fig.cap="Figure 16: try out themes"}
p16 <- ggplot(variantData[variantData$type=='SNP',])+
  geom_bar(aes(x = trv_type, fill = tier))+
  facet_grid(variant~reference)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5), legend.position = 'top', axis.title.x = element_blank())
p16
```

## changing the order of aesthetic mappings

```{r plot20, echo=FALSE, fig.cap="Figure 20: re-ordering stuff"}
variantData$reference <- factor(variantData$reference, levels = rev(levels(variantData$reference)))
variantData$trv_type <- factor(variantData$trv_type, levels = rev(levels(variantData$trv_type)))

p20 <- ggplot(data = variantData[variantData$type=='SNP',])+
  geom_bar(aes(x = trv_type, fill = tier))+
  facet_grid(variant~reference)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1, size = 5),
        legend.position = 'top', axis.title.x = element_blank())
p20
```

## Saving ggplot2 plots

```{r how to safe (gg)plots, echo=TRUE}
pdf(file = 'p20.pdf', height = 8, width = 11)
p20
dev.off()

ggsave('p20_2.pdf', device = 'pdf')
```

## Wide vs. long foramt

```{r wide vs. long, echo=TRUE}
ggplot(Orange, aes(x = age, y = circumference))+
  geom_point()

library(reshape2)

Orange2 <- melt(data=Orange, id.vars=c("Tree"))
ggplot(Orange2, aes(x=value, fill=variable))+
  geom_density()

```

# ggplot2 Practice examples

```{r load data2, include=FALSE}
tt <- read.delim("http://genomedata.org/gen-viz-workshop/intro_to_ggplot2/ggplot2ExampleData.tsv")
tt <- tt[tt$dataset == "discovery",]
```

```{r P1, echo=FALSE, fig.cap="Fig.1) Violins"}
library(plyr)
variantDataCount <- count(tt, "Simple_name")
variantDataMax <- aggregate(data=variantData, tumor_VAF ~ Simple_name, max)
iii <- merge(variantDataMax, variantDataCount)
#head(iii)

P1 <- ggplot()+
  geom_violin(data = tt, aes(x = Simple_name, y = tumor_VAF, fill = Simple_name))+
  theme_classic()+
  geom_point(data = tt, aes(x = Simple_name, y = tumor_VAF), position = 'jitter',
             size = .5, shape = 4)+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 5))+
  xlab('Sample')+
  ylab('VAF')+
  guides(fill=FALSE)+
  geom_text(data = iii, aes(x = Simple_name, y = tumor_VAF + 5, label = freq))
P1
```

